backend/hidrive/hidrive.go

```diff
@@ -0,0 +1,1021 @@
+		// Object was not found, so create a new one.
+		return f.PutUnchecked(ctx, in, src, options...)
+	}
+	return nil, err
+}
+
+// PutStream uploads the contents of the io.Reader to the remote path
+// with the modTime given of indeterminate size.
+// The existing or new object is returned.
+//
+// A new object may have been created or
+// an existing one accessed even if an error is returned,
+// in which case both the object and the error will be returned.
+func (f *Fs) PutStream(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	fs.Debugf(f, "executing PutStream(%s, %v).", src.Remote(), options)
+
+	return f.Put(ctx, in, src, options...)
+}
+
+// PutUnchecked the contents of the io.Reader into the remote path
+// with the modTime given of the given size.
+// This guarantees that existing objects will not be overwritten.
+// The new object is returned.
+//
+// This will produce an error if an object already exists at that path.
+//
+// In case the upload fails and an object has been created,
+// this will try to delete the object at that path.
+// In case the failed upload could not be deleted,
+// both the object and the (upload-)error will be returned.
+func (f *Fs) PutUnchecked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	remote := src.Remote()
+	modTime := src.ModTime(ctx)
+	fs.Debugf(f, "executing PutUnchecked(%s, %v).", remote, options)
+	resolvedPath := f.resolvePath(remote)
+
+	// NOTE: The file creation operation is a single atomic operation.
+	// Thus uploading as much content as is reasonable
+	// (i.e. everything up to the cutoff) in the first request,
+	// avoids files being created on upload failure for small files.
+	// (As opposed to creating an empty file and then uploading the content.)
+	tmpReader, bytesRead, err := readerForChunk(in, int(f.opt.UploadCutoff))
+	cutoffReader := cachedReader(tmpReader)
+	if err != nil {
+		return nil, err
+	}
+
+	var info *api.HiDriveObject
+	err = f.retrier.Call(func() (bool, error) {
+		var createErr error
+		// Reset the reading index (in case this is a retry).
+		if _, createErr = cutoffReader.Seek(0, io.SeekStart); createErr != nil {
+			return false, createErr
+		}
+		info, createErr = f.createFile(ctx, resolvedPath, cutoffReader, modTime, IgnoreOnExist)
+
+		if createErr == fs.ErrorDirNotFound {
+			// Create the parent-directory for the object and repeat request.
+			_, parentErr := f.createDirectories(ctx, path.Dir(resolvedPath), IgnoreOnExist)
+			if parentErr != nil && parentErr != fs.ErrorDirExists {
+				fs.Errorf(f, "Tried to create parent-directory for '%s', but failed.", resolvedPath)
+				return false, parentErr
+			}
+			return true, createErr
+		}
+		return false, createErr
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	o, err := f.newObjectFromHiDriveObject(remote, info)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if bytesRead < int(f.opt.UploadCutoff) {
```

Comment:
Integer overflow? uploadcutoff could be bigger than 2GB which on a 32bit machine will cause integer overflow.

Reply:
Hopefully it never is, because as the help indicates, 2GiB is the maximum HiDrive can handle.
But overflows are ugly regardless, will fix this.